### ISSUE TITLE
403 Permission error when using SPN auth type

### ISSUE THREAD

**Issue by francky-powerfactors** [Azure/azure-storage-fuse#587]
### Which version of the blobfuse was used?
1.3.6

### Which OS (please include version) are you using?
Linux minikube 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 GNU/Linux

### What problem was encountered?
I always get a 403 when blobfuse is trying to authenticate with the storage account. My service principal has the `Contributor` role assigned to it for the storage account. It also has the `Storage Blob Data Contributor` and the `Storage Queue Data Contributor` roles but I still can't authenticate with this service principal.

In the container it tries to connect to (in the overview menu) I see that the 'Authentication method' is 'Access key'. Should it be 'Azure AD User Account' instead?

**My hope is that someone can tell me exactly what permissions / roles the service principal I'm using must have in order to successfully authenticate with the storage account.**

There are probably tests for this in your codebase, how are your permissions / roles setup?

### Have you found a mitigation/solution?
If I use a SAS token it works. But this add a lot of complexity since, I have to create / manage the token and inject it into our Kubernetes setup so that the correct values are put into the secret.

I would rather want to use a service principal.

### By default, blobfuse logs errors to syslog. If this is relevant, is there anything in the syslog that might be helpful?
```
Mar 26 21:06:08 minikube blobfuse[1093]: Account name found
Mar 26 21:06:08 minikube blobfuse[1093]: Container name found
Mar 26 21:06:08 minikube blobfuse[1093]: Auth type set as spn
Mar 26 21:06:08 minikube blobfuse[1093]: Setting logging level to : LOG_DEBUG
Mar 26 21:06:08 minikube blobfuse[1093]: release is 5.400000, Kernel version is 5.4.72-microsoft-standard-WSL2
Mar 26 21:06:08 minikube blobfuse[1093]: CURL Version is 7.64.0
Mar 26 21:06:08 minikube blobfuse[1093]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making cache directory /mnt.
Mar 26 21:06:08 minikube blobfuse[1093]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making cache directory /mnt/resource.
Mar 26 21:06:08 minikube blobfuse[1093]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making cache directory /mnt/resource/blobfusetmp.
Mar 26 21:06:08 minikube blobfuse[1093]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making cache directory /mnt/resource/blobfusetmp/root.
Mar 26 21:06:08 minikube blobfuse[1093]: Initializing blobfuse with attr caching enabled
Mar 26 21:06:08 minikube blobfuse[1093]: Initializing blobfuse using BlockBlob
Mar 26 21:06:08 minikube blobfuse[1093]: Kernel version is 5.400000
Mar 26 21:06:08 minikube blobfuse[1093]: libcurl version is 7.640000
Mar 26 21:06:08 minikube blobfuse[1093]: Authenticating using SPN
Mar 26 20:52:11 minikube blobfuse[1087]: SPN auth uri token request url = https://login.microsoftonline.com/<tenant id>/oauth2/v2.0/token
Mar 26 20:52:11 minikube blobfuse[1087]: refresh callback set
Mar 26 20:52:11 minikube blobfuse[1087]: calling refresh token
Mar 26 20:52:17 minikube blobfuse[1087]:  ==> REQUEST/RESPONSE :: POST https://login.microsoftonline.com/<tenant id>/oauth2/v2.0/token?&Content-Type=application/x-www-form-urlencoded&Content-Length=174&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 200 :: REQ ID : c7c114c1-9f2c-4e3c-8369-6c50475e3802
Mar 26 20:52:17 minikube blobfuse[1087]: INFO not WARNING: Printing Json Token as string for visibility{"access_token":"<access token>","expires_in":3599,"ext_expires_in":3599,"token_type":"Bearer"}
Mar 26 20:52:17 minikube blobfuse[1087]: INFO: After adding specified expires_in token expiry time in utc Fri Mar 26 21:52:16 2021#012
Mar 26 20:52:17 minikube blobfuse[1087]: OAUTH Token : Refresh token succeeded expiry time is: Fri Mar 26 21:52:16 2021
Mar 26 20:52:17 minikube blobfuse[1087]:  ==> REQUEST/RESPONSE :: GET https://pipelinepipe01metast01.blob.core.windows.net/metadata?comp=list&delimiter=/&include=metadata&maxresults=1&restype=container?&User-Agent=azure-storage-fuse/1.3.6&x-ms-date=Fri, 26 Mar 2021 20:52:17 GMT&x-ms-version=2018-11-09&Authorization=****&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 403 :: REQ ID : 31f1585e-d01e-006a-13ff-228b41000000
Mar 26 20:52:17 minikube blobfuse[1087]: Unable to start blobfuse.  Failed to connect to the storage container. There might be something wrong about the storage config, please double check the storage account name, account key/sas token/OAuth access token and container name. errno = 403
Mar 26 20:52:17 minikube blobfuse[1087]: Unable to start blobfuse due to a lack of credentials. Please check the readme for valid auth setups.
```

This is my config file:
```
accountName pipelinepipe01metast01
containerName metadata
servicePrincipalClientId <client id> => application id of the app registration
servicePrincipalTenantId <tenant id>
servicePrincipalClientSecret <client secret>
authType SPN
```

### If relevant, please share your mount command.
`blobfuse /blobfuse-test --tmp-path=/mnt/resource/blobfusetmp --config-file=/blobfuse-config/config.cfg --use-https=true --file-cache-timeout-in-seconds=120 --log-level=LOG_DEBUG --use-attr-cache=true`

**Comment by NaraVen** [comment]
I see that you are using WSL. It is still in beta mode and we have not tested it with SPN.

**Comment by francky-powerfactors** [comment]
Ok.

Using a SAS token works correctly but using a service principal or using the account storage key doesn't work. For those 2 cases, I always get a 403.

Is it something that you (in the larger sense) plan to support in the near future? I'm trying to deploy my app in minikube the exact same way as we would normally deploy it in AKS but this issue is blocking me from having the exact same setup running in both environment.

Thank you for your quick reply.

**Comment by NaraVen** [comment]
Hmmm storage Key itself is failing, Is your clocktime synced correctly? 
Also looks like the bearer token you are getting shows an expired time. After adding specified expires_in token expiry time in utc Fri Mar 26 21:52:16 2021#012 . The request is being made at Mar 26 21:52:17 that's why it is giving 403.

**Comment by francky-powerfactors** [comment]
What do you mean by _The request is being made at Mar 26 21:52:17_?

I'm getting a refresh token at 20:52:11 here
`Mar 26 20:52:11 minikube blobfuse[1087]: calling refresh token`, which is valid for 1h and then the lib makes a request to Azure here
`Mar 26 20:52:17 minikube blobfuse[1087]:  ==> REQUEST/RESPONSE :: GET https://pipelinepipe01metast01.blob.core.windows.net...`

at Mar 26 20:52:17, which is well before Mar 26 21:52:16 2021

Am I reading something wrong?

**Comment by NaraVen** [comment]
what is the time on the Minikube ? Is it 15 minutes behind original time by any chance?
Mar 26 

Messages with  21:06:08  are preceeding messages with 20:52 timestamp. Any known reason for that?
See below extract of the logs


minikube blobfuse[1093]: Authenticating using SPN
Mar 26 20:52:11 minikube blobfuse[1087]: SPN auth uri token request url = https://login.microsoftonline.com/<tenant id>/oauth2/v2.0/token
Mar 26 20:52:11 minikube blobfuse[1087]: refresh callback set

**Comment by francky-powerfactors** [comment]
Right, I get what you mean.

This was probably me copying logs once, and then I wanted to add more details so I copied the first logs from another attempt. I'll try and post a better version of the logs where all the times makes sense.

**Comment by francky-powerfactors** [comment]
Here a version where all the timestamps are in order

```
Mar 30 00:40:58 minikube blobfuse[851]: Account name found
Mar 30 00:40:58 minikube blobfuse[851]: Container name found
Mar 30 00:40:58 minikube blobfuse[851]: Auth type set as spn
Mar 30 00:40:58 minikube blobfuse[851]: Setting logging level to : LOG_DEBUG
Mar 30 00:40:58 minikube blobfuse[851]: release is 5.400000, Kernel version is 5.4.72-microsoft-standard-WSL2
Mar 30 00:40:58 minikube blobfuse[851]: CURL Version is 7.64.0
Mar 30 00:40:58 minikube blobfuse[851]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making c
ache directory /mnt.
Mar 30 00:40:58 minikube blobfuse[851]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making c
ache directory /mnt/resource.
Mar 30 00:40:58 minikube blobfuse[851]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making c
ache directory /mnt/resource/blobfusetmp.
Mar 30 00:40:58 minikube blobfuse[851]: Function ensure_files_directory_exists_in_cache, in file /home/vsts/work/1/s/blobfuse/utilities.cpp, line 96: Making c
ache directory /mnt/resource/blobfusetmp/root.
Mar 30 00:40:58 minikube blobfuse[851]: Initializing blobfuse with attr caching enabled
Mar 30 00:40:58 minikube blobfuse[851]: Initializing blobfuse using BlockBlob
Mar 30 00:40:58 minikube blobfuse[851]: Kernel version is 5.400000
Mar 30 00:40:58 minikube blobfuse[851]: libcurl version is 7.640000
Mar 30 00:40:58 minikube blobfuse[851]: Authenticating using SPN
ache directory /mnt/resource/blobfusetmp/root.
Mar 30 00:40:58 minikube blobfuse[851]: Initializing blobfuse with attr caching enabled
Mar 30 00:40:58 minikube blobfuse[851]: Initializing blobfuse using BlockBlob
Mar 30 00:40:58 minikube blobfuse[851]: Kernel version is 5.400000
Mar 30 00:40:58 minikube blobfuse[851]: libcurl version is 7.640000Mar 30 00:40:58 minikube blobfuse[851]: Authenticating using SPN
Mar 30 00:40:58 minikube blobfuse[851]: SPN auth uri token request url = https://login.microsoftonline.com/<tenant id>/oauth2/v2.0/token
Mar 30 00:40:58 minikube blobfuse[851]: refresh callback set
Mar 30 00:40:58 minikube blobfuse[851]: calling refresh token
Mar 30 00:41:03 minikube blobfuse[851]:  ==> REQUEST/RESPONSE :: POST https://login.microsoftonline.com/<tenant id>/oauth2/v2.0/token?&Content-Type=application/x-www-form-urlencoded&Content-Length=174&Tra
nsfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 200 :: REQ ID : acfb9fd3-0306-492f-a821-48ad7fcfad00Mar 30 00:41:03 minikube blobfuse[851]: INFO not WARNING: Printing Json Token as string for visibility{"access_token":"<access token>","expires_in":3599,
"ext_expires_in":3599,"token_type":"Bearer"}Mar 30 00:41:03 minikube blobfuse[851]: INFO: After adding specified expires_in token expiry time in utc Tue Mar 30 01:41:02 2021#012
Mar 30 00:41:03 minikube blobfuse[851]: OAUTH Token : Refresh token succeeded expiry time is: Tue Mar 30 01:41:02 2021Mar 30 00:41:04 minikube blobfuse[851]:  ==> REQUEST/RESPONSE :: GET https://pipelinepipe01metast01.blob.core.windows.net/metadata?comp=list&delimiter=/&include=metadata&maxresults=1&restype=container?&User-Agent=azure-storage-fu
se/1.3.6&x-ms-date=Tue, 30 Mar 2021 00:41:03 GMT&x-ms-version=2018-11-09&Authorization=****&Transfer-Encoding=--------------------------------------------------------------------------------RESPONSE Status :: 403 :: REQ ID : d574
ff9b-601e-0040-2489-255451000000Mar 30 00:41:04 minikube blobfuse[851]: Unable to start blobfuse.  Failed to connect to the storage container. There might be something wrong about the storage config, please double check the storage account name, account key/sas
 token/OAuth access token and container name. errno = 403
Mar 30 00:41:04 minikube blobfuse[851]: Unable to start blobfuse due to a lack of credentials. Please check the readme for valid auth setups.
```

But I think you were right about the time of the container. I'm did this request at 13h23 on the 30th, GTM-4 (EST with daylight saving) but the container is saying the current date is:

```
# date
Tue Mar 30 00:44:50 UTC 2021
```

which doesn't make sense... I'll look into that.

**Comment by francky-powerfactors** [comment]
OMG you were 100% right... I searched and found https://github.com/microsoft/WSL/issues/4114. I ran the command `sudo hwclock -s` in WSL and the time automatically updated in minikube. I ran the blobfuse command again and guess what...

```
Mar 30 17:39:43 minikube blobfuse[856]: Authenticating using SPN
...
Mar 30 17:39:49 minikube blobfuse[856]: Successfully Authenticated!
```

Thanks for all your help @NaraVen!

For all the others having this issue, look at your clock to make sure it's synched with you're host's clock!
